src/backend/executor/execTuples.c

```diff
@@ -1596,6 +1596,14 @@ slot_getsysattr(TupleTableSlot *slot, int attnum, Datum *result, bool *isnull)
                         case TableOidAttributeNumber:
 							*result = ObjectIdGetDatum(slot->tts_tableOid);
 							break;
+						case MinTransactionIdAttributeNumber:
+						case MaxTransactionIdAttributeNumber:
+						case MinCommandIdAttributeNumber:
+						case MaxCommandIdAttributeNumber:
+							if (slot->PRIVATE_tts_flags & TTS_APPEND_OPTIMIZED) {
+								elog(ERROR, "AO-table has no xmin, xmax, cmin, cmax, attributes");
```

Comment:
Why we still need TTS_APPEND_OPTIMIZED and error on execute?

Reply:
It will work if tuple opened with the path skipping SQL parser, f.e. directly from some code. I agree that it sounds a bit fantastic, and it looks like double check.

Reply:
If this error message is necessary, please add a test which will print it

Reply:
This message will be printed when table with attributes created by previous version opened with this version. It can happen after upgrade with existing tables. AFAIK in the current tests we can not do it.